docs(grpc): document gRPC API and protocol

npepinpe opened the pull request:
closes #1669

npepinpe commented:
@wints I requested your review, mostly because of the `grpc/README.md` page; you don't need to read the reference page. You wrote a post about writing a gRPC client which I reference, and I thought maybe we should at some point add something in there. I also kind of ripped off the post for this, hopefully plagerizing is fine ;)

wints commented:
@npepinpe Looks good to me! And yes, of course OK to take content from blog posts. Feel free to do this for docs at any point.

Zelldon commented on docs/src/grpc/reference.md:
```diff
@@ -0,0 +1,681 @@
+
+Returned if:
+
+  - the job was marked as failed. In that case, the related incident must be resolved before
```
Or it is not activated?

npepinpe replied:
At the moment we allow a job to be completed even if not activated. We can discuss that.

Zelldon commented on docs/src/grpc/reference.md:
```diff
@@ -0,0 +1,681 @@
+  // the assigned key, which acts as a unique identifier for this workflow
+  int64 workflowKey = 3;
+  // the resource name (see: WorkflowRequestObject.name) from which this workflow was
+  / /parsed
```
```suggestion
  // parsed
```

Zelldon commented on docs/src/grpc/reference.md:
```diff
@@ -0,0 +1,681 @@
+Returned if:
+
+  - no workflows have been deployed yet (if no bpmnProcessId was given)
+  - no workflow with the given process ID exists (if bpmnProcessId was given)
```
Why not just returning empty list :sweat_smile:

Zelldon commented on gateway-protocol/src/main/proto/gateway.proto:
```diff
@@ -5,222 +5,395 @@ option java_multiple_files = false;
-  ResourceType type = 2;
-  bytes definition = 3;
-}
+// ActiveJobs
```
These kind of comments are really verbose/redundant :sweat_smile:

npepinpe replied:
It was mostly to group things together, I think maybe we should think about splitting into multiple files and having the `gateway.proto` be simply the service definition? But I can remove them now, they aren't useful.

Zelldon commented on docs/src/grpc/reference.md:
```diff
@@ -0,0 +1,681 @@
+  // the job type, as defined in the BPMN process (e.g. <zeebe:taskDefinition
+  // type="payment-service" />)
+  string type = 1;
+  // the name of the worker activating the jobs, mostly used for logging purposes
```
Ok I'm old. the worker name is not checked anymore on completing the job :see_no_evil:

npepinpe commented on docs/src/grpc/README.md:
```diff
@@ -0,0 +1,27 @@
+
+Zeebe uses gRPC for external client to server communication, and clients connect to brokers via a stateless gateway.
+The client protocol is defined using Protocol Buffers v3 (proto3), which you can find in the
+[main Zeebe repository](https://github.com/zeebe-io/zeebe/tree/develop/gateway-protocol)
```
Hehe, this is the part I took directly from Mike's blog, so I didn't actually write that ;)

wints replied:
Heh, sometimes I don't quite get the wording right--I'd be fine with changing to something like what you proposed, @Zelldon

npepinpe commented on docs/src/grpc/README.md:
```diff
@@ -0,0 +1,27 @@
+If Zeebe does not provide an officially-supported client in your target language, you can read the official [Quick Start](https://grpc.io/docs/quickstart/) page to find out how
+to create a very basic one.
+
+In addition to this, a [blog post](https://zeebe.io/blog/2018/11/grpc-generating-a-zeebe-python-client/) was already published with a short tutorial on how to write one for Python.
```
As in, a list of clients? Maybe I could just link to the [list of awesome zeebe clients](https://github.com/zeebe-io/awesome-zeebe#clients)?

npepinpe commented:
> Nice that we finally have this now. Might be better if we import the Gateway protocol, as we do in the code examples or zeebe config - with that we do not have to maintain both.

What do you mean? I kind of like having it formatted and split into different calls so you can use the TOC, but I admit it's more effort. If we split the protocol into multiple files though we could import only the related definitions for the input/output stuff, which could be nice. But since I don't expect we modify this so often at this point, we could leave this for the next iteration.

> Do we plan to update also the basics section https://docs.zeebe.io/basics/client-server.html , here the gateway should be mentioned and explained right?

There is an issue for this #1538 which I think Mike offered to write with Menski (not sure what's the status on this), we can bring it up at the next team meeting.

And I collected the points about the protocol you raised in #1910, we can discuss them tomorrow or next week. I would say for now we just document them.

Zelldon commented:
I meant that we can directly import the protobuf file like we do here https://docs.zeebe.io/java-client-examples/workflow-deploy.html . Because otherwise we have to make sure that both are up to date. The protobuf file can be self-explanatory, like you already did.

I wouldn't split the protocol file, it makes it unecessary harder to generate the clients.

npepinpe commented:
> I wouldn't split the protocol file, it makes it unecessary harder to generate the clients.

Doesn't the compiler just load a folder? At least with the maven plugin you just point it to the folder and it'll parse everything, no?

I like the idea of linking the file, but then we'd need to ensure at least:

1. We have correct syntax highlighting
1. We need to somehow document the errors more thoroughly in the file

Or did you mean to link it as a separate page, and not a replacement for the current reference page? Because the reference page I wrote has two advantages: it has a table of contents, and more thorough documentation (esp. the errors).

wints commented:
Indeed, I started working on #1538 today and should have a version ready for review by the end of this week.

Zelldon commented:
Ok I thought you have to specify each file, but yes could work.

Yes the first one, as a replacement.
Lets discuss this tomorrow maybe @menski has an opinion on this?

Zelldon commented:
>  it has a table of contents, and more thorough documentation (esp. the errors).

Agree, maybe we can have the sections and only import parts of the file? (I think we have to impl this)

But anyway I would like to maintain only one file.

menski commented:
> Agree, maybe we can have the sections and only import parts of the file? (I think we have to impl this)

Ranged includes are supported by mdbook
https://rust-lang-nursery.github.io/mdBook/format/mdbook.html#including-files

> But anyway I would like to maintain only one file.

This is not 100% true as you still have to maintain the range includes in the documentation when you change the proto file. If you remember that's why you decided to include the whole Zeebe configuration file instead of ranges, when you documented the cluster settings. But maybe it's still less effort to maintain only the ranges.

Zelldon commented:
Yes you are right :D

npepinpe commented:
So what's the end result? It's not extremely important to me one way or another if we have the formatted page or just include the file. So if you guys prefer just including the file then we'll go with that.

menski commented:
> I pushed an example of how it could look like. Personally I like the formatted one better, but I don't feel that strongly about it so as you guys prefer.

I prefer the version before this change. Let's go with it for now and adjust if we figure out it is to much effort.

npepinpe commented:
bors r+
